misc-utils/lsclocks.c

```diff
@@ -0,0 +1,366 @@
+static void scols_line_asprintf(struct libscols_line *ln, size_t n, const char *format, ...)
+{
+	char *data;
+	va_list args;
+
+	va_start(args, format);
+	xvasprintf(&data, format, args);
+	va_end(args);
+
+	scols_line_refer_data(ln, n, data);
+}
+
+static void scols_line_format_timespec(struct libscols_line *ln, size_t n, const struct timespec *ts)
+{
+	scols_line_asprintf(ln, n, "%ju.%09" PRId32, (uintmax_t) ts->tv_sec, (uint32_t) ts->tv_nsec);
+}
+
+static clockid_t parse_clock(const char *name)
+{
+	size_t i;
+	uint32_t id = -1;
+	int rc;
+
+	rc = ul_strtou32(name, &id, 10);
+
+	for (i = 0; i < ARRAY_SIZE(clocks); i++) {
+		if (!strcmp(name, clocks[i].id_name)
+		    || !strcmp(name, clocks[i].name))
+			return clocks[i].id;
+		if (rc == 0 && (clockid_t) id == clocks[i].id)
+			return id;
+	}
+
+	errx(EXIT_FAILURE, _("Unknown clock: %s"), name);
+}
+
+int main(int argc, char **argv)
+{
+	size_t i, j;
+	char c;
+	int rc;
+	const struct colinfo *colinfo;
+	const struct clockinfo *clockinfo;
+
+	struct libscols_table *tb;
+	struct libscols_line *ln;
+	struct libscols_column *col;
+
+	bool noheadings = false, raw = false, json = false;
+	const char *outarg = NULL;
+	int columns[ARRAY_SIZE(infos)];
+	size_t ncolumns = 0;
+	clockid_t clock = -1;
+
+	struct timespec resolution, now;
+	char buf[FORMAT_TIMESTAMP_MAX];
+
+	static const struct option longopts[] = {
+		{ "noheadings", no_argument,       NULL, 'n' },
+		{ "output",     required_argument, NULL, 'o' },
+		{ "version",    no_argument,       NULL, 'V' },
+		{ "help",	no_argument,       NULL, 'h' },
+		{ "json",       no_argument,       NULL, 'J' },
+		{ "raw",        no_argument,       NULL, 'r' },
+		{ "time",       required_argument, NULL, 't' },
+		{ 0 },
+	};
+
+	setlocale(LC_ALL, "");
+	bindtextdomain(PACKAGE, LOCALEDIR);
+	textdomain(PACKAGE);
+	close_stdout_atexit();
+
+	while ((c = getopt_long(argc, argv, "no:JrVh", longopts, NULL)) != -1) {
+		switch (c) {
+		case 'n':
+			noheadings = true;
+			break;
+		case 'o':
+			outarg = optarg;
+			break;
+		case 'J':
+			json = true;
+			break;
+		case 'r':
+			raw = true;
+			break;
+		case 't':
+			clock = parse_clock(optarg);
+			break;
+		case 'V':
+			print_version(EXIT_SUCCESS);
+		case 'h':
+			usage();
+		default:
+			errtryhelp(EXIT_FAILURE);
+		}
+	}
+
+	if (argv[optind])
+		errtryhelp(EXIT_FAILURE);
+
+	if (clock != -1) {
+		rc = clock_gettime(clock, &now);
+		if (rc)
+			err(EXIT_FAILURE, _("failed to get time"));
+		printf("%ju.%09"PRId32"\n", (uintmax_t) now.tv_sec, (uint32_t) now.tv_nsec);
+		return EXIT_SUCCESS;
+	}
+
+	if (outarg && string_add_to_idarray(outarg, columns, ARRAY_SIZE(columns),
+					    &ncolumns, column_name_to_id) < 0)
+		return EXIT_FAILURE;
+
+	if (!ncolumns) {
+		columns[ncolumns++] = COL_ID;
+		columns[ncolumns++] = COL_NAME;
+		columns[ncolumns++] = COL_HUMAN_NAME;
+		columns[ncolumns++] = COL_TIME;
+		columns[ncolumns++] = COL_ISO_TIME;
+		columns[ncolumns++] = COL_RESOLUTION;
+	}
```

Comment:
Please, move this default before string_add_to_idarray(). The function string_add_to_idarray() can extend the default if you use `-o +colname` or it can completely replace the content of the array if you use `-o colname`.